### DISCUSSION TITLE
"Not a valid process webhook job class"

### DISCUSSION THREAD

**Discussion by hicles** [spatie/laravel-webhook-client#178] [Q&A] [answered]
Hi, I've copied 1:1 code from the documentation to handle a webhook but it's throwing this error:
```"`App\\Jobs\\ProcessWebhookJob` is not a valid process webhook job class. A valid class should implement `Spatie\\WebhookClient\\Jobs\\ProcessWebhookJob`.```

Handler implementation
```
<?php

namespace App\Jobs;

use Spatie\WebhookClient\Jobs\ProcessWebhookJob as SpatieProcessWebhookJob;

class ProcessWebhookJob extends SpatieProcessWebhookJob
{
    public function handle()
    {
        logger('works');
        // $this->webhookCall // contains an instance of `WebhookCall`

        // perform the work here
    }
}
```

Config implementation:
```        
/*
    * The class name of the job that will process the webhook request.
    *
    * This should be set to a class that extends \Spatie\WebhookClient\Jobs\ProcessWebhookJob.
    */
'process_webhook_job' => App\Jobs\ProcessWebhookJob::class
```

**Comment by hicles** [comment, marked as answer]
Unsure why this was moved to Q&A since it clearly seems to be an issue with the package itself. I wasn't able to figure out what was causing this or solve it so I just decided to make my own webhooks handler instead of using this package.

**Reply by freekmurze** [reply]
Our tests indicate that our package is working as intended. Probably it didn't work for you due to misconfiguration.

Should you give the package another shot and find that the package is indeed in error, PR a failing test for us to fix.

**Comment by iRajul** [comment]
I guess name of file should be `ProcessWebhookJob.php`

**Comment by QuentinGab** [comment]
For anybody having the same `is not a valid process webhook job class` error:
You must specify `process_webhook_job` on EVERY item of the `configs`, the **default value of `process_webhook_job` in the 'default' config is empty**, you MUST set something or delete completely the default config otherwise it will raise an error everytime !

**Reply by Richard2012** [reply]
Is it OK to have in the config
`    'name' => 'default',`
and
`    `'process_webhook_job' => App\Http\Controllers\Hooks\HookJob::class,`` ?

I have run into the issue and when checking the code where **process_webhook_job** config var is imported into the app (last line of WebhookConfig.php) I got the following warning from my IDE:
![a](https://github.com/user-attachments/assets/1e1ef815-27c2-441d-8eb8-c697196b9ba2)

In my config I have
`'process_webhook_job' => App\Http\Controllers\Hooks\HookJob::class,`
and yes, my class does extend \Spatie\WebhookClient\Jobs\ProcessWebhookJob as requested.

Our problem is that our HookJob::class never gets called.
The headers and the payload are written into the DB and the error logs are empty.

Would you have a comment on this? Thanks.

**Comment by chrisk-7777** [comment]
Short answer: `php artisan queue:restart` as part of your deployment process.

Long answer:

I've noticed this issue pop up in a few discussions, I also hit the same issue.  
For us, it was because ProcessWebhookJob is a job, and jobs live on queues that are long lived (unless its set to `sync` which is unlikely in production environments).  

We had our queue already running under supervisor (via Forge), we added this library, but the process part of the flow broke because it wasn't "aware" of this new code. 

As per the Laravel docs: 

> Since queue workers are long-lived processes, they will not notice changes to your code without being restarted

https://laravel.com/docs/10.x/queues#queue-workers-and-deployment


This is particularly tricky to trace because (for us) it only presented itself in production.
